File: Sprint6/K/C++/main.cpp
// https://contest.yandex.ru/contest/25069/problems/K/
#include <iostream>
#include <vector>
#include <utility>
#include <climits>

using namespace std;

void dijkstra(int vertex_s, vector<vector<pair<int, int>>> &graph, vector<int> &dist) {
    vector<bool> visited(graph.size(), false);
    dist[vertex_s] = 0;

    for (int i = 0; i < graph.size(); ++i) {
        int u = -1;

        for (int j = 0; j < graph.size(); ++j) {
            if (!visited[j] && (u == -1 || dist[j] < dist[u])) {
                u = j;
            }
        }

        if (dist[u] == INT_MAX) {
            break;
        }

        visited[u] = true;

        for (auto &edge : graph[u]) {
            int v = edge.first;
            int w = edge.second;

            if (dist[u] + w < dist[v]) {
                dist[v] = dist[u] + w;
            }
        }
    }
}

int main() {
    int n, m;
    cin >> n >> m;

    vector<vector<std::pair<int, int>>> graph(n);

    for (int i = 0; i < m; ++i) {
        int u, v, l;
        cin >> u >> v >> l;
        graph[u - 1].emplace_back(v - 1, l);
        graph[v - 1].emplace_back(u - 1, l);
    }

    for (int i = 0; i < n; ++i) {
        vector<int> dist(n, INT_MAX);
        dijkstra(i, graph, dist);

        for (int j = 0; j < n; ++j) {
            if (dist[j] == INT_MAX) {
                cout << -1 << ' ';
            } else {
                cout << dist[j] << ' ';
            }
        }
        cout << endl;
    }

    return 0;
}

File: Sprint6/K/Rust/src/main.rs
// https://contest.yandex.ru/contest/25069/problems/K/

use std::collections::BinaryHeap;
use std::collections::HashSet;
use std::io::{self, BufRead};

#[derive(Eq, PartialEq)]
struct State {
    cost: i32,
    vertex: usize,
}

impl Ord for State {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        other.cost.cmp(&self.cost)
    }
}

impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

fn dijkstra(vertex_s: usize, graph: &[Vec<(usize, i32)>], dist: &mut [i32]) {
    let mut visited = HashSet::new();
    let mut heap = BinaryHeap::new();

    heap.push(State {
        cost: 0,
        vertex: vertex_s,
    });
    dist[vertex_s] = 0;

    while let Some(State { cost, vertex }) = heap.pop() {
        if visited.contains(&vertex) {
            continue;
        }

        visited.insert(vertex);

        for &(next_vertex, weight) in &graph[vertex] {
            let next_cost = cost + weight;

            if next_cost < dist[next_vertex] {
                heap.push(State {
                    cost: next_cost,
                    vertex: next_vertex,
                });
                dist[next_vertex] = next_cost;
            }
        }
    }
}

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let first_line = lines.next().unwrap().unwrap();
    let nm: Vec<usize> = first_line
        .split_whitespace()
        .map(|x| x.parse().unwrap())
        .collect();
    let (n, m) = (nm[0], nm[1]);

    let mut graph: Vec<Vec<(usize, i32)>> = vec![vec![]; n];

    for _ in 0..m {
        let edge: Vec<i32> = lines
            .next()
            .unwrap()
            .unwrap()
            .split_whitespace()
            .map(|x| x.parse::<i32>().unwrap())
            .collect();
        let (u, v, l) = (edge[0] as usize, edge[1] as usize, edge[2]);
        graph[u - 1].push((v - 1, l));
        graph[v - 1].push((u - 1, l));
    }

    for i in 0..n {
        let mut dist = vec![i32::MAX; n];
        dijkstra(i, &graph, &mut dist);

        for d in &dist {
            if *d == i32::MAX {
                print!("-1 ");
            } else {
                print!("{} ", d);
            }
        }
        println!();
    }
}
